libsolidity/analysis/ControlFlowRevertPruner.h

```diff
@@ -0,0 +1,149 @@
+
+	/// Analyze all member functions of the contract
+	/// including inherited and overriden functions from base contracts
+	bool visit(ContractDefinition const& _contract)
+	{
+		for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts)
+			for (FunctionDefinition const* function: contract->definedFunctions())
+				removeRevertingPaths(*function, &_contract);
+
+		return false;
+	}
+
+	/// Analyze free functions
+	bool visit(FunctionDefinition const& _function)
+	{
+		if (_function.isFree())
+			removeRevertingPaths(_function, nullptr);
+
+		return false;
+	}
+
+	/// Removes any paths that are still pending after the first iteration
+	void removePendingPaths();
+
+	/// Find nodes that did function calls that will revert and remove any exits
+	/// from them
+	void removeRevertingPaths(FunctionDefinition const& _function, ContractDefinition const* _contract);
+
+	/// Recursively analyze a given function call for reverts, but don't
+	/// resolve any recursive loops.
+	/// @params _functionCall function call to analyze.
+	/// @params _contract Contract from which the call is made.
+	/// @returns the revert state of the function call.
+	RevertState checkForReverts(FunctionCall const& _functionCall, ContractDefinition const* _contract);
+
+	/// Recursively traverses function flows, analyzing the nodes for possible
+	/// reverts.
+	/// @param _function function definition whose flow will be traversed.
+	/// @param _contract contract that called that function.
+	/// @param _flow function flow to be traversed.
+	/// @param _onRevertState delegate to modify behavior based on the
+	///        revert state of a function call.
+	///        If it returns false, processing of that node will stop.
+	/// @returns the revert state of the requested function call.
+	RevertState traverseFunctionFlow(
+		FunctionDefinition const& _function,
+		ContractDefinition const* _contract,
+		FunctionFlow const& _flow,
+		std::function<bool (RevertState, CFGNode*, FunctionFlow const&)> _onRevertState
+	);
+
+	/// @returns a key object for the function flow / function revert containers
+	/// @param _function function definition to build the key for
+	/// @param _contract contract definition from which the call is made
+	std::tuple<ContractDefinition const*, FunctionDefinition const*> getNormalizedKey(FunctionDefinition const& _function, ContractDefinition const* _contract);
+
+	// Remembers the states of already processed functions.
+	// Anything left pending will be processed in a second step.
+	std::map<
+		std::tuple<ContractDefinition const*, FunctionDefinition const*>,
+		std::tuple<CFGNode*, RevertState>
+	> m_functionReverts;
+
+	// Control Flow Graph object.
+	CFG const& m_cfg;
+};
+
+
+/// Resolves a function call using
+/// - the calling contract
+/// - the unresolved function definition
+/// - the function call expression
+///
+class ResolveFunction: private ASTConstVisitor
```

Comment:
This is needlessly complicated. Please write a function similar to `ASTNode::referencedDeclaration` that performs the lookup based on an `Expression` only - you don't need to pass in the FunctionDefinition, do you?